### PR TITLE
fix(display): check data after reload from self-service

### PR DESCRIPTION
**Pull request by stonebuzz**
When self-service select an itil categorie from tick

form is reload and fields don't take care of new value

Only work for itilcategories_id because thatt trigger a reload of form

Fix : (portal) !6633

### CODE REVIEW

File: inc/containerdisplaycondition.class.php [hunk 1]
```diff
@@ -362,7 +362,7 @@ public static function removeBlackListedOption($array, $itemtype_class)
     }
 
 
-    public function computeDisplayContainer($item, $container_id)
+    public function computeDisplayContainer($item, $container_id, $params = null)
```
Comment (cedric-anne) [inline inc/containerdisplaycondition.class.php:365]: 1. `$input` seems to be a more appropriate name.
2. It should always be passed as an array.
```suggestion
    public function computeDisplayContainer($item, $container_id, array $input)
```

File: inc/containerdisplaycondition.class.php [hunk 2]
```diff
@@ -372,7 +372,7 @@ public function computeDisplayContainer($item, $container_id)
             $display = true;
             foreach ($found_dc as $data) {
                 $displayCondition->getFromDB($data['id']);
-                $result = $displayCondition->checkCondition($item);
+                $result = $displayCondition->checkCondition($item, $params);
```
Comment (cedric-anne) [inline inc/containerdisplaycondition.class.php:375]:
```suggestion
                $result = $displayCondition->checkCondition($item, $input);
```

File: inc/containerdisplaycondition.class.php [hunk 3]
```diff
@@ -386,42 +386,49 @@ public function computeDisplayContainer($item, $container_id)
     }
 
 
-    public function checkCondition($item)
+    public function checkCondition($item, $params = null)
```
Comment (cedric-anne) [inline inc/containerdisplaycondition.class.php:389]:
```suggestion
    public function checkCondition($item, array $input)
```

File: inc/containerdisplaycondition.class.php [hunk 4]
```diff
@@ -386,42 +386,49 @@ public function computeDisplayContainer($item, $container_id)
...
         $value = $this->fields['value'];
         $condition = $this->fields['condition'];
         $searchOption = Search::getOptions(get_class($item))[$this->fields['search_option']];
 
+        //if from self-service form
+        if ($params != null) {
+            $object_fields = $params;
+        } else {
+            $object_fields = $item->fields;
+        }
```
Comment (cedric-anne) [inline inc/containerdisplaycondition.class.php:400]:
```suggestion
        $object_fields = array_merge($item->fields, $input);
```

File: inc/field.class.php [hunk 5]
```diff
@@ -750,8 +750,45 @@ public static function showForTab($params)
             return false;
         }
 
+        //JS to trigger any change and check if container need to be display or not
+        $ajax_url   = Plugin::getWebDir('fields') . '/ajax/container_display_condition.php';
+        $items_id = $item->getID() ? $item->getID() : 0;
```
Comment (cedric-anne) [inline inc/field.class.php:755]:
```suggestion
        $items_id = !$item->isNewItem() ? $item->getID() : 0;
```

File: inc/field.class.php [hunk 6]
```diff
@@ -750,8 +750,45 @@ public static function showForTab($params)
...
+        echo Html::scriptBlock(<<<JAVASCRIPT
+
+            function checkContainerVisibility(){
+                var data = $('#itil-form').serializeArray().reduce(function(obj, item) {
```
Comment (cedric-anne) [inline inc/field.class.php:759]: `#itil-form` will make this only work on ITIL objects. Many other forms can be targetted by `.asset_form`, but, in fact, we cannnot be sure that all forms could be targetted this way.
Maybe the selector should be `$('#fields_container').closest('form')`.

Also, having a random value appended to identifier may prevent issues on pages that may display multiple forms (ITIL timeline for instance).

File: inc/containerdisplaycondition.class.php (hunk 2 above)
Comment (stonebuzz) [reply inc/containerdisplaycondition.class.php]: done

File: inc/containerdisplaycondition.class.php (hunk 3 above)
Comment (stonebuzz) [reply inc/containerdisplaycondition.class.php]: done

File: inc/containerdisplaycondition.class.php (hunk 4 above)
Comment (stonebuzz) [reply inc/containerdisplaycondition.class.php]: done

**Review by cedric-anne — APPROVED** [review]

**Review by AdrienClairembault — APPROVED** [review]

**Pull request merged**